### PR TITLE
Fix mobile composer input regressions

### PR DESCRIPTION
**Pull request by klopez4212**
## Summary

- reuse the channel member snapshot so first-use `@` suggestions appear immediately
- reduce selection-only composer rebuilds so iOS selection handles stay responsive
- make Return insert a newline, send only from the composer button, and animate multiline growth with reduced-motion support

## Validation

- `just mobile-check`
- `just mobile-test` — 1,271 tests passed
- signed iPhone Release and Pixel 10 debug builds installed and launched
- `just ci` passed mobile, Rust, desktop, and web checks until the unrelated `buzz-terminal` lifecycle test timed out waiting for `$0`; reproduced unchanged in isolation

### CODE REVIEW

**Comment by klopez4212** [comment]
@codex review

**Review by wesbillman — COMMENTED** [review]
Reviewed exact head `c971857a150da6a62de0a5be26bc7026a55b0708` on Wes's behalf. No blocking findings.

The state transitions are coherent: composer rebuilds now select only `controller.text`; text/selection replacement is atomic through `TextEditingValue`; typing and draft persistence are gated on actual text changes; rich-text caching includes every color baked into the returned spans and is invalidated when mention labels change; composing ranges remain part of the cache key. Return/newline behavior is explicit and button-only send remains covered.

The first-use mention path also holds together: channel-list membership events seed immutable per-channel snapshots, the latest event wins, relay changes clear the cache, and both the provider and profile preload path use the snapshot only until the authoritative member request resolves. The connection-aware member provider re-runs when session status changes.

Tests intentionally change and pin user-visible behavior: Return now inserts a newline instead of sending, multiline height animates unless reduced motion is enabled, and the iOS magnifier is disabled while selection handles remain enabled. The last item trades the native loupe for responsive selection; it appears deliberate and is non-blocking, but it is worth a real-device smoke check.

Exact-head evidence: Mobile CI and all reported checks are green; `git diff --check origin/main...origin/pr-5594` passes; GitHub reports the branch mergeable. I did not duplicate the full mobile suite locally.

**Comment by brow** [comment]
🤖 Reviewed at `c971857a150da6a62de0a5be26bc7026a55b0708`. One blocker, otherwise this looks good and the intent is clearly right.

**Blocker: member mention suggestions disappear during a reconnect.**

`channelMembersProvider` now returns an empty list when the session is not connected. `SessionState` is watched as a whole, so a `connected` to `reconnecting` transition re-runs the provider, the new guard returns early, and the value becomes `AsyncData([])`.

That empty list is not null, so it wins over the new snapshot in both consumers:

- `ComposeBar` uses `membersAsync.asData?.value ?? cachedMembers`
- `mentionCandidatesProvider` uses `loadedMembers ?? cachedMembersForChannel(...)`

The result is that the snapshot this PR adds cannot do its job in exactly the moment it is most useful. Measured in a widget test with the real provider and the real popover, with a snapshot available throughout:

| state | at `c971857a15` | at merge base `cf03bd7c37` |
| --- | --- | --- |
| connected | suggestions shown | suggestions shown |
| reconnecting | **popover absent** | suggestions shown |
| reconnected | suggestions shown | suggestions shown |

This is reachable in normal use, not only on a dropped socket. `onAppResumed` calls `_connect` when the app was backgrounded for at least the 5 second grace period, and `_connect` sets `reconnecting` once the session has connected before. So backgrounding the app for a few seconds and returning puts the composer in this state while the socket re-establishes. The channel and thread composers stay mounted across that window. It does recover after reconnect, and no wrong content is ever sent.

Two options, either is fine:

1. Do not publish an authoritative empty value while disconnected. Keep the previous value or stay in a state the consumers treat as cache-eligible.
2. Have both consumers treat a disconnected empty result as cache-eligible, so the snapshot fills the gap.

A regression test across connected, reconnecting, and connected would pin it. The current provider test covers query suppression and the eventual refetch, but not what the composer shows in the middle state.

**Non-blocking notes**

- Watching all of `SessionState` also re-runs the provider on `reconnectAttempt` changes during retry backoff. No extra relay fetches result, because the guard returns first, and the only `connected` write is a canonical const with attempt zero. Worth knowing, not worth changing here.
- `ChannelMember.joinedAt` in the snapshot is the event's `createdAt`, so it is the snapshot time rather than each member's real join time. Nothing in mobile reads `joinedAt` today, so this is not user-visible, but it is a small trap for the next reader.
- The new composer invariants have no regression detector in the test suite. We deleted the new `useEffect` listener guard entirely, then each half of it separately, then the draft dedupe early return, and the full mobile suite stayed green each time (1271 passed). The suite can fail, so this is a measured gap and not an unobservable one: inverting `canSend` reds it immediately. The tests in `compose_bar_test.dart` pin the visible pairings, such as multiline with newline and the single-notification splice, rather than these guards. A future simplification of the guard would go unnoticed.
- The `buildTextSpan` cache key omits `colors.primary` while the agent mention chip uses it. This is safe as written: the chip is a `WidgetSpan` child that reads the theme at its own build time, so an accent change repaints it even when the cached span is reused. Verified, no change needed. Calling it out only so the omission does not look accidental later.

Everything else in the diff held up under review. The atomic `TextEditingValue` update, the text-change gating for typing and draft persistence, the composing range in the cache key, and Return inserting a newline with send staying on the button are all coherent and covered by the new tests. CI is green at job and step level at this head.

**Comment by klopez4212** [comment]
@codex review

**Comment by klopez4212** [comment]
@codex review

**Comment by klopez4212** [comment]
@codex review

File: mobile/lib/features/channels/channel_management_provider.dart [hunk 1]
```diff
@@ -421,7 +441,9 @@ final channelDetailsProvider = FutureProvider.family<ChannelDetails, String>((
 final channelMembersProvider = FutureProvider.autoDispose
     .family<List<ChannelMember>, String>((ref, channelId) async {
       ref.watch(channelMembershipUpdateProvider(channelId));
-      final session = ref.watch(relaySessionProvider.notifier);
+      final sessionState = ref.watch(relaySessionProvider);
+      if (sessionState.status != SessionStatus.connected) return const [];
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply mobile/lib/features/channels/channel_management_provider.dart]: Fixed in `09467b7a0` and verified again at `2f5212fa8`. `channelMembersProvider` no longer publishes an authoritative empty value while connecting/reconnecting: it retains the current provider value, uses a relay/account/channel-scoped snapshot for mounts during the reconnect window, and invalidates only on the transition back to connected. This protects member sheets, permissions, and mention consumers without changing composer behavior or UI.

File: mobile/lib/features/channels/channels_provider.dart [hunk 2]
```diff
@@ -62,7 +72,11 @@ class ChannelsNotifier extends AsyncNotifier<List<Channel>> {
 
   @override
   Future<List<Channel>> build() async {
-    ref.watch(relayConfigProvider);
+    final relayBaseUrl = ref.watch(relayConfigProvider).baseUrl;
+    if (_memberSnapshotRelayBaseUrl != relayBaseUrl) {
+      _memberSnapshotRelayBaseUrl = relayBaseUrl;
+      _memberSnapshotsByChannelId = const {};
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply mobile/lib/features/channels/channels_provider.dart:82]: Fixed in `09467b7a0`/`feb215cd8` and verified at `2f5212fa8`. The channel-list member snapshot cache now resets when either the relay base URL or active pubkey changes, so replacing credentials for the same relay cannot reuse the previous identity’s member snapshots.

File: mobile/lib/features/channels/channel_management_provider.dart [hunk 3]
```diff
@@ -421,17 +484,47 @@ final channelDetailsProvider = FutureProvider.family<ChannelDetails, String>((
...
+          ref.invalidateSelf();
+        }
+      });
+      final sessionState = ref.read(relaySessionProvider);
+      if (sessionState.status != SessionStatus.connected) {
+        return snapshotCache.read(
+          relayBaseUrl: relayBaseUrl,
+          pubkey: pubkey,
+          channelId: channelId,
+        );
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply mobile/lib/features/channels/channel_management_provider.dart]: Fixed in `feb215cd8` and covered by the new regression test in `2f5212fa8`. When the provider first mounts during reconnect and its own snapshot cache is empty, it seeds from `ChannelsNotifier.cachedMembersForChannel`; the test confirms the retained members are returned and no disconnected member fetch is issued.

**Comment by klopez4212** [comment]
@codex review

**Review by brow — APPROVED** [review]

**Comment by brow** [comment]
🤖 Re-reviewed at `2f5212fa8a3b546fa7d95a39019a60a7c42d30ca`. **The reconnect blocker is fixed.** Approving.

Verified with the same instrument that found the bug, run at three revisions so the result has a control: the merge base, the previously reviewed head, and this head. During a connected to reconnecting transition, the previous head settled on an authoritative empty member list and the mention popover disappeared. This head keeps the members visible at every pump depth, so there is no window at all, not even a single frame. Recovery also works: returning to connected re-fetches, and it still re-fetches after a fetch that threw.

Four non-blocking notes. None of them should hold this PR, and the last three are all much smaller than the bug you just fixed.

**1. A test for the recovery half.** Your new provider test goes connected to reconnecting and stops there, so the part that re-fetches after the reconnect completes has no coverage. We checked by narrowing the guard that triggers the refresh to a condition that only breaks recovery: the full mobile suite still passed 1274 of 1274. The behavior you shipped is correct, but one more test that goes connected, reconnecting, connected and asserts the fetch count went up would protect it.

**2. A superseded fetch can write stale members into the new cache.** The cache write happens after the await, and a replaced provider run is not cancelled, so if something invalidates the provider while a fetch is in flight, the older run finishes last and its write wins. Measured with a control: the live value was correctly `[alice]` after a member was removed, while the cache held the pre-removal `[alice, bob]`. The visible value stays correct, so this only surfaces if the widget remounts during a later disconnected window, where autocomplete could offer a member who was removed. It clears on the next successful fetch. Moving the cache write so it only applies when the run is still current would close it.

**3. An empty result from a healthy fetch is cached and then shadows the channel-list snapshot.** When a connected fetch returns no member events, the code caches the empty list. Later, while offline, the cache read returns that empty list, which is not null, so the function returns before it can fall back to the channel-list snapshot. Worth knowing that all the real failure modes are safe here: a timeout, a relay-side close, and a dropped connection all complete with an error rather than an empty list, so only a genuinely empty channel reaches this path. That makes the cached entry useless rather than wrong.

**4. A whole-object session watch in the composer.** `ComposeBar` now watches the entire session state to read the status, in a PR whose purpose is to reduce composer rebuilds. That state object has no equality override and gains a new instance on each reconnect backoff attempt with only the attempt counter changing, so five retries produce six composer builds where one would do. Selecting just the status is a one-line change. We confirmed this stays inside the composer's own builds: the attempt-only churn causes no extra member fetches at any of the three revisions.

Also worth stating plainly, because it was the obvious thing to worry about: a member removed while the app was offline can still be resolved when sending during a reconnect window. That behavior is the same on the merge base, so this PR does not introduce it, and the relay remains the authority on delivery.

For visibility on what these measurements do not cover: everything above is provider-level and widget-level testing, so none of it speaks to how long a real reconnect lasts on a device or to real socket behavior. CI is green at this head at the job and step level, which is a claim about those jobs and not a re-run of the suite, though we did also run the full mobile suite locally at this head: 1275 passed, none failed.
